### PR TITLE
Add default L2 network offerings

### PR DESCRIPTION
**Pull request by nvazquez**
## Description
Add default L2 network offerings

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
image

image

## Checklist:
- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 1]
```diff
@@ -561,6 +561,35 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
                     offering.setDedicatedLB(false);
                     _networkOfferingDao.update(offering.getId(), offering);
                 }
+
+                if (_networkOfferingDao.findByUniqueName(NetworkOffering.DefaultL2NetworkOfferingNoVlan) == null) {
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java]: can you create a method for addDefaultL2Offerings() ?

File: server/src/main/java/com/cloud/server/ConfigurationServerImpl.java [hunk 2]
```diff
@@ -1175,6 +1175,31 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
                     _ntwkOfferingServiceMapDao.persist(offService);
                     s_logger.trace("Added service for the network offering: " + offService);
                 }
+
+                NetworkOfferingVO l2NoVlan = new NetworkOfferingVO(NetworkOffering.DefaultL2NetworkOfferingNoVlan,
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/ConfigurationServerImpl.java]: can you create a addDefaultL2Offerings(), please?

**Comment by nvazquez** [comment]
Thanks @DaanHoogland, I have added new methods for them

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
i'll wait with merging for your answer on the duplication question, @nvazquez. I see no way in which this is touched by any integration test, but can you assess them anyway?

**Comment by nvazquez** [comment]
Thanks @DaanHoogland and @rhtyd, I've refactored it to avoid code duplication.
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: engine/schema/src/com/cloud/offerings/dao/NetworkOfferingDaoImpl.java [hunk 3]
```diff
@@ -221,4 +224,41 @@ public boolean isUsingServicePackage(String uuid) {
 
         return false;
     }
+
+    @Override
+    public void persistDefaultL2NetworkOfferings() {
```
[the remark this answers is not in the document]
Comment (nvazquez) [reply engine/schema/src/com/cloud/offerings/dao/NetworkOfferingDaoImpl.java]: There are 2 occurrences: one as you mention for new environments (init global setting = false) and the other on NetworkOrchestrator.configure(). You are right, to support the second case we should check if the offerings already exists before adding them, I'll refactor it

**Comment by PaulAngus** [comment]
A build in Trillian failed this morning as the management server couldn't start.  WIth the following error:

2018-06-06 07:32:13,639 DEBUG [c.c.c.ConfigurationManagerImpl] (main:null) (logid:) Adding network offering [Network Offering [0-Guest-QuickCloudNoServices]
2018-06-06 07:32:13,675 DEBUG [c.c.u.d.T.Transaction] (main:null) (logid:) Rolling back the transaction: Time = 40 Name =  tx-13; called by -TransactionLegacy.rollback:889-TransactionLegacy.removeUpTo:832-TransactionLegacy.close:656-TransactionContextInterceptor.invoke:36-ReflectiveMethodInvocation.proceed:174-ExposeInvocationInterceptor.invoke:92-ReflectiveMethodInvocation.proceed:185-JdkDynamicAopProxy.invoke:212-$Proxy63.persist:-1-NetworkOfferingDaoImpl.persistDefaultL2NetworkOfferings:259-NativeMethodAccessorImpl.invoke0:-2-NativeMethodAccessorImpl.invoke:62
2018-06-06 07:32:13,676 WARN  [o.a.c.s.m.c.ResourceApplicationContext] (main:null) (logid:) Exception encountered during context initialization - cancelling refresh attempt: org.springframework.context.ApplicationContextException: Failed to start bean 'cloudStackLifeCycle'; nested exception is com.cloud.utils.exception.CloudRuntimeException: DB Exception on: com.mysql.jdbc.PreparedStatement@23038621: INSERT INTO ntwk_offering_service_map (ntwk_offering_service_map.network_offering_id, ntwk_offering_service_map.service, ntwk_offering_service_map.provider, ntwk_offering_service_map.created) VALUES (0, _binary'UserData', _binary'ConfigDrive', '2018-06-06 07:32:13')
2018-06-06 07:32:13,677 WARN  [o.e.j.w.WebAppContext] (main:null) (logid:) Failed startup of context o.e.j.w.WebAppContext@365c30cc{/client,file:///usr/share/cloudstack-management/webapp/,UNAVAILABLE}{/usr/share/cloudstack-management/webapp}
org.springframework.context.ApplicationContextException: Failed to start bean 'cloudStackLifeCycle'; nested exception is com.cloud.utils.exception.CloudRuntimeException: DB Exception on: com.mysql.jdbc.PreparedStatement@23038621: INSERT INTO ntwk_offering_service_map (ntwk_offering_service_map.network_offering_id, ntwk_offering_service_map.service, ntwk_offering_service_map.provider, ntwk_offering_service_map.created) VALUES (0, _binary'UserData', _binary'ConfigDrive', '2018-06-06 07:32:13')
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:186)
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:52)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:358)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:159)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:123)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:884)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:552)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.loadContext(DefaultModuleDefinitionSet.java:145)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet$2.with(DefaultModuleDefinitionSet.java:122)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.withModule(DefaultModuleDefinitionSet.java:245)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.withModule(DefaultModuleDefinitionSet.java:250)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.withModule(DefaultModuleDefinitionSet.java:250)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.withModule(DefaultModuleDefinitionSet.java:233)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.loadContexts(DefaultModuleDefinitionSet.java:117)
	at org.apache.cloudstack.spring.module.model.impl.DefaultModuleDefinitionSet.load(DefaultModuleDefinitionSet.java:79)
	at org.apache.cloudstack.spring.module.factory.ModuleBasedContextFactory.loadModules(ModuleBasedContextFactory.java:37)
	at org.apache.cloudstack.spring.module.factory.CloudStackSpringContext.init(CloudStackSpringContext.java:71)
	at org.apache.cloudstack.spring.module.factory.CloudStackSpringContext.<init>(CloudStackSpringContext.java:58)
	at org.apache.cloudstack.spring.module.factory.CloudStackSpringContext.<init>(CloudStackSpringContext.java:62)
	at org.apache.cloudstack.spring.module.web.CloudStackContextLoaderListener.contextInitialized(CloudStackContextLoaderListener.java:52)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:890)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:532)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:853)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:344)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1515)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1477)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:785)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:261)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:545)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:133)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:107)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.doStart(GzipHandler.java:273)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:133)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:115)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:133)
	at org.eclipse.jetty.server.Server.start(Server.java:418)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:107)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.server.Server.doStart(Server.java:385)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.apache.cloudstack.ServerDaemon.start(ServerDaemon.java:200)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.commons.daemon.support.DaemonLoader.start(DaemonLoader.java:241)
Caused by: com.cloud.utils.exception.CloudRuntimeException: DB Exception on: com.mysql.jdbc.PreparedStatement@23038621: INSERT INTO ntwk_offering_service_map (ntwk_offering_service_map.network_offering_id, ntwk_offering_service_map.service, ntwk_offering_service_map.provider, ntwk_offering_service_map.created) VALUES (0, _binary'UserData', _binary'ConfigDrive', '2018-06-06 07:32:13')
	at com.cloud.utils.db.GenericDaoBase.persist(GenericDaoBase.java:1436)
	at com.cloud.offerings.dao.NetworkOfferingServiceMapDaoImpl.persist(NetworkOfferingServiceMapDaoImpl.java:169)
	at com.cloud.offerings.dao.NetworkOfferingServiceMapDaoImpl.persist(NetworkOfferingServiceMapDaoImpl.java:34)
	... 87 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Cannot add or update a child row: a foreign key constraint fails (`cloud`.`ntwk_offering_service_map`, CONSTRAINT `fk_ntwk_offering_service_map__network_offering_id` FOREIGN KEY (`network_offering_id`) REFERENCES `network_offerings` (`id`) ON DELETE CASCADE)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	... 105 more
2018-06-06 07:32:13,680 INFO  [o.e.j.s.h.ContextHandler] (main:null) (logid:) Started o.e.j.s.h.MovedContextHandler@80169cf{/,null,AVAILABLE}
2018-06-06 07:32:13,722 INFO  [o.e.j.s.AbstractConnector] (main:null) (logid:) Started ServerConnector@15d0c81b{HTTP/1.1,[http/1.1]}{:::8080}
2018-06-06 07:32:13,723 INFO  [o.e.j.s.Server] (main:null) (logid:) Started @73420ms

**Comment by nvazquez** [comment]
@PaulAngus yes, that's related to what @rhtyd mentioned, I've added a fix for that
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Based on code reviews and tests LGTM. I also see that it now adds the default offerings only if they don't exist already. Merging this now.

**Pull request merged**